src/ChainSafe.Gaming.MetaMask.Unity/MetaMaskController.cs

```
@@ -0,0 +1,197 @@
+
+                // Even if request fails we still need to callback so request task completion source result can be set.
+                onResponse?.Invoke(default);
+
+                throw rpcRequest.Exception;
+            }
+
+            logger.Log($"Successful {data.Method} JsonRPC response with result {rpcRequest.Result}");
+
+            onResponse?.Invoke(rpcRequest.Result);
+        }
+
+        /// <summary>
+        /// Callback when MetaMask successfully connects to an account.
+        /// </summary>
+        /// <param name="address">Connected Address.</param>
+        public void EthereumEnabled(string address)
+        {
+            logger.Log("Ethereum Enabled.");
+
+            if (!isInitialized)
+            {
+                // Subscribe to callbacks.
+                MetamaskWebglInterop.EthereumInit(gameObject.name, nameof(NewAccountSelected), nameof(ChainChanged));
+
+                // Get and log ChainId.
+                MetamaskWebglInterop.GetChainId(gameObject.name, nameof(ChainChanged), nameof(DisplayError));
+
+                isInitialized = true;
+            }
+
+            InvokeAccountConnected(address);
+        }
+
+        private void InvokeAccountConnected(string address)
+        {
+            OnAccountConnected?.Invoke(address);
+        }
+
+        /// <summary>
+        /// Callback when a new account is selected by user.
+        /// </summary>
+        /// <param name="address">New selected account address.</param>
+        public void NewAccountSelected(string address)
+        {
+            logger.Log($"New Account with address {address} Selected.");
+
+            ConnectedAddress = address;
+        }
+
+        /// <summary>
+        /// Callback when chain is changed on Network.
+        /// </summary>
+        /// <param name="chainId">New chain.</param>
+        public void ChainChanged(string chainId)
```

We can't allow chain changing also.

---

So Metamask allows it and we can't inhibit Metamasks function, what we can do it maybe throw an exception or log an error when chain is changed

---

Yup, do that 😁

---

done